### PR TITLE
[framework] Emits coin::CreatedCurrency metadata + compiles book examples

### PR DESCRIPTION
**Pull request by damirka**
- Implements suggestion from #4551. Now call to `create_currency` emits `coin::CreatedCurrency` event.
- Adds code from examples book to framework tests (this way we make sure they always compile)

Closes #4551

### CODE REVIEW

File: crates/sui-framework/sources/coin.move [hunk 1]
```diff
@@ -33,6 +34,15 @@ module sui::coin {
         total_supply: Supply<T>
     }
 
+    // === Events ===
+
+    /// Emitted when new currency is created through the `create_currency` call.
+    /// Contains currency metadata for off-chain discovery. Type parameter `T`
+    /// matches the one in `Coin<T>`
+    struct CurrencyCreated<phantom T> has copy, drop {
+        decimals: u8
```
Comment (sblackshear) [inline crates/sui-framework/sources/coin.move:43]:
```suggestion
        /// Number of decimal places the coin uses. 
        /// A coin with `value ` N and `decimals` D should be shown as N / 10^D
        /// E.g., a coin with `value` 7002  and decimals 3 should be displayed as 7.002
        /// This is metadata for display usage only.
        decimals: u8
```

porting over comments from the other PR

**Review by sblackshear — COMMENTED** [review]
Shouldn't this need to specify the `decimals` for `SUI` somewhere?

**Comment by damirka** [comment]
> Shouldn't this need to specify the `decimals` for `SUI` somewhere?

Nice catch. SUI is created through balance and we don’t have a TC for it, balance::Supply<SUI> is stored inside the SuiSystem. 

I wonder if we should eventually instantiate SUI with an OTW but that’s def gonna be tricky.

**Comment by damirka** [comment]
This change introduces a whole set of questions related to balances (as in not Coins). 

But I believe that we should keep these functions around Coins while keeping the balance module. Happy to hear your thoughts @sblackshear!

**Comment by sblackshear** [comment]
I didn't realize that there is no TC for Sui--I think it's useful to have one + put it inside `SuiSystem` in case there are future governance proposals that require (e.g.) burning Sui. 

In terms of coin vs balance vs supply, I must admit I still don't have a great understanding of when a programmer would use each of these three (probably a good topic for a future doc on coins and fungible tokens). But in the absence of other info, I would imagine that `decimals` and other token metadata would be relevant for all three of these.

**Review by sblackshear — CHANGES_REQUESTED** [review]
until we can figure out how to support decimals for SUI

**Comment by damirka** [comment]
@sblackshear since we know that sui::SUI is a special case in verifier and can be constructed outside initializer, I’ll submit a patch to return TC instead of Supply in a couple days.

**Comment by damirka** [comment]
@sblackshear Sorry for the delay. SUI is now initialized through `coin::create_currency`.

File: crates/sui-framework/sources/coin.move [hunk 2]
```diff
@@ -33,6 +34,19 @@ module sui::coin {
...
 
+    // === Events ===
+
+    /// Emitted when new currency is created through the `create_currency` call.
+    /// Contains currency metadata for off-chain discovery. Type parameter `T`
+    /// matches the one in `Coin<T>`
+    struct CurrencyCreated<phantom T> has copy, drop {
+        /// Number of decimal places the coin uses. 
+        /// A coin with `value ` N and `decimals` D should be shown as N / 10^D
+        /// E.g., a coin with `value` 7002  and decimals 3 should be displayed as 7.002
```
Comment (sblackshear) [inline crates/sui-framework/sources/coin.move:45]:
```suggestion
        /// E.g., a coin with `value` 7002 and decimals 3 should be displayed as 7.002
```

File: crates/sui-framework/sources/sui.move [hunk 3]
```diff
@@ -3,19 +3,22 @@
...
     struct SUI has drop {}
 
-    /// Register the token to acquire its `TreasuryCap`.
+    /// Register the `SUI` Coin to acquire its `Supply`.
     /// This should be called only once during genesis creation.
-    public(friend) fun new(): Supply<SUI> {
-        balance::create_supply(SUI {})
+    public(friend) fun new(ctx: &mut TxContext): Supply<SUI> {
+        coin::treasury_into_supply(
+            coin::create_currency(SUI {}, 6, ctx)
```
Comment (sblackshear) [inline crates/sui-framework/sources/sui.move:20]:
```suggestion
            coin::create_currency(SUI {}, 9, ctx)
```

The minimum unit is 10^-9 SUI, which we call a "mist" (i.e., 1 nano-SUI = 1 mist). I think we should also have a comment explaining this, perhaps on a `DECIMALS` constant so this information is easier to discover for someone reading the module?

**Review by sblackshear — CHANGES_REQUESTED** [review]

**Review by sblackshear — APPROVED** [review]

**Pull request merged**